File: Github.dart
import 'package:http/http.dart' as http;
import 'package:args/args.dart';
import 'dart:convert';
import 'dart:io';

const BASE_URL = "https://api.github.com/users/";

class GthUserItem {
  int count;
  var arr = <String>[];

  GthUserItem(this.count);
}

class GthUser {
  String username;

  late String name;
  late String bio;
  late String link;
  late GthUserItem repos;
  late GthUserItem gists;
  late GthUserItem followers;
  late GthUserItem following;

  GthUser(this.username);

  static Future<GthUser> create(String user) async {
    final self = GthUser(user);
    var res = await http.get(Uri.parse(BASE_URL + self.username));
    var obj = jsonDecode(res.body);

    self.name = obj["name"];
    self.bio = obj["bio"];
    self.link = obj["html_url"];

    self.repos = GthUserItem(obj["public_repos"]);
    self.gists = GthUserItem(obj["public_gists"]);
    self.followers = GthUserItem(obj["followers"]);
    self.following = GthUserItem(obj["following"]);

    return self;
  }

  Future<void> fetch(String thing) async {
    var url = "${BASE_URL + this.username}/$thing";
    var res = await http.get(Uri.parse(url));
    var arr = jsonDecode(res.body);

    switch (thing) {
      case "repos":
        for (var v in arr)
          this.repos.arr.add(v["name"]);

        break;

      case "gists":
        for (var v in arr)
          this.gists.arr.add(v["description"]);

        break;

      case "followers":
        for (var v in arr)
          this.followers.arr.add(v["login"]);

        break;

      case "following":
        for (var v in arr)
          this.following.arr.add(v["login"]);

        break;

      default:
        break;
    }
  }
}

void showHelp() {
  print("Github [OPTIONS...] <USERNAMES...>");
  print("");
  print("Options:");
  print("  -r, --repos      Show user repos                                     [boolean]");
  print("  -g, --gists      Show user gists                                     [boolean]");
  print("  -f, --followers  Show user followers                                 [boolean]");
  print("  -F, --following  Show user following                                 [boolean]");
  print("  -h, --help       Show this help                                      [boolean]");
}

void main(final List<String> args) async {
  final parser = ArgParser();
  parser.addFlag("repos", abbr: "r", defaultsTo: false);
  parser.addFlag("gists", abbr: "g", defaultsTo: false);
  parser.addFlag("followers", abbr: "f", defaultsTo: false);
  parser.addFlag("following", abbr: "F", defaultsTo: false);
  parser.addFlag("help", abbr: "h", defaultsTo: false);

  final opts = parser.parse(args);

  if (opts["help"]) {
    showHelp();
    exit(0);
  }

  if (args.isEmpty) {
    print("Nothing to do.");
    print("Pass -h or --help to see the help.");
    exit(1);
  }

  for (final arg in opts.rest) {
    final user = await GthUser.create(arg);

    print("Name: ${user.name}");
    print("Bio: ${user.bio}");
    print("Link: ${user.link}");

    print("Public repos: ${user.repos.count}");
    if (opts["repos"]) {
      await user.fetch("repos");

      for (final r in user.repos.arr)
        print(" | $r");
    }

    print("Public gists: ${user.gists.count}");
    if (opts["gists"]) {
      await user.fetch("gists");

      for (final g in user.gists.arr)
        print(" | $g");
    }

    print("Followers: ${user.followers.count}");
    if (opts["followers"]) {
      await user.fetch("followers");

      for (final u in user.followers.arr)
        print(" | @$u");
    }

    print("Following: ${user.following.count}");
    if (opts["following"]) {
      await user.fetch("following");

      for (final u in user.following.arr)
        print(" | @$u");
    }

    print("");
  }
}

File: Github.ts
import yargs from "https://deno.land/x/yargs@v17.4.0-deno/deno.ts";

// Yeah I'm lazy af
const print = console.log;
const BASE_URL = "https://api.github.com/users/";

type GthUserItem = {
  count: number;
  arr: string[];
};

class GthUser {
  name: string = "";
  bio: string = "";
  link: string = "";
  repos: GthUserItem = { count: 0, arr: [] };
  gists: GthUserItem = { count: 0, arr: [] };
  followers: GthUserItem = { count: 0, arr: [] };
  following: GthUserItem = { count: 0, arr: [] };

  constructor(public username: string) {}

  static async create(user: string): Promise<GthUser> {
    let self = new GthUser(user);
    let url = BASE_URL + self.username;
    let res = await fetch(url);
    
    if (!res.ok)
      throw new Error(`can't fetch url '${url}'`);

    let obj = await res.json();

    self.name = obj.name;
    self.bio  = obj.bio;
    self.link = obj.html_url;

    self.repos.count = obj.public_repos;
    self.gists.count = obj.public_gists;
    self.followers.count = obj.followers;
    self.following.count = obj.following;

    return self;
  }

  async fetch(thing: string) {
    let url = `${BASE_URL + this.username}/${thing}`
    let res = await fetch(url);

    if (!res.ok)
      throw new Error(`can't fetch url '${url}'`);

    let arr = await res.json();

    switch (thing) {
      case "repos":
        for (const val of arr)
          this.repos?.arr.push(val.name);
        break;

      case "gists":
        for (const val of arr)
          this.gists?.arr.push(val.description);
        break;

      case "followers":
        for (const val of arr)
          this.followers?.arr.push(val.login);
        break;

      case "following":
        for (const val of arr)
          this.following?.arr.push(val.login);
        break;

      default:
        throw new Error(`unsupported endpoint: ${thing}`);
    }
  }
}

let opts = yargs(Deno.args)
  .version(false)
  .usage("Github [options...] <usernames...>")
  .option("repos", {
    alias: "r",
    type: "boolean",
    describe: "Show user repos"
  })
  .option("gists", {
    alias: "g",
    type: "boolean",
    describe: "Show user gists"
  })
  .option("followers", {
    alias: "f",
    type: "boolean",
    describe: "Show user followers"
  })
  .option("following", {
    alias: "F",
    type: "boolean",
    describe: "Show user following"
  })
  .help("help", "Show this help")
  .alias("h", "help")
  .parse();

if (opts._.length <= 0) {
  print("Nothing to do.");
  print("Pass -h or --help to see the help.");
  Deno.exit(1);
}

for (const val of opts._) {
  let user = await GthUser.create(val);

  print(`Name: ${user.name}`);
  print(`Bio: ${user.bio}`);
  print(`Link: ${user.link}`);

  print(`Public repos: ${user.repos?.count}`);
  if (opts.r) {
    await user.fetch("repos");

    if (user.repos)
      for (const idx in user.repos.arr)
        print(` | ${parseInt(idx) + 1} ${user.repos.arr[parseInt(idx)]}`);
  }

  print(`Public gists: ${user.gists?.count}`);
  if (opts.g) {
    await user.fetch("gists");

    if (user.gists)
      for (const idx in user.gists.arr)
        print(` | ${parseInt(idx) + 1} ${user.gists.arr[parseInt(idx)]}`);
  }

  print(`Public followers: ${user.followers?.count}`);
  if (opts.followers) {
    await user.fetch("followers");

    if (user.followers)
      for (const val of user.followers.arr)
        print(` | @${val}`);
  }

  print(`Public following: ${user.following?.count}`);
  if (opts.following) {
    await user.fetch("following");

    if (user.following)
      for (const val of user.following.arr)
        print(` | @${val}`);
  }

  print();
}